nuxeo-core/nuxeo-core-storage/src/main/java/org/nuxeo/ecm/core/storage/FulltextExtractorWork.java

```diff
@@ -88,141 +118,207 @@ public String getTitle() {
-        DocumentLocation docLocation = new DocumentLocationImpl(doc);
-        List<IndexAndText> indexesAndText = new LinkedList<>();
+    protected void extractAndUpdateBinaryText() {
+        // we extract binary text even if fulltext search is disabled,
+        // because it is still used to inject into external indexers like Elasticsearch
+        BlobsExtractor blobsExtractor = new BlobsExtractor();
+        Map<Blob, String> blobsText = new IdentityHashMap<>();
         for (String indexName : fulltextConfiguration.indexNames) {
             if (!fulltextConfiguration.indexesAllBinary.contains(indexName)
                     && fulltextConfiguration.propPathsByIndexBinary.get(indexName) == null) {
                 // nothing to do: index not configured for blob
                 continue;
             }
-            extractor.setExtractorProperties(fulltextConfiguration.propPathsByIndexBinary.get(indexName),
+            // get original text from all blobs
+            blobsExtractor.setExtractorProperties(fulltextConfiguration.propPathsByIndexBinary.get(indexName),
                     fulltextConfiguration.propPathsExcludedByIndexBinary.get(indexName),
                     fulltextConfiguration.indexesAllBinary.contains(indexName));
-            List<Blob> blobs = extractor.getBlobs(doc);
-            StringBlob stringBlob = blobsToStringBlob(blobs, docId);
-            String text = fulltextParser.parse(stringBlob.getString(), null, stringBlob.getMimeType(), docLocation);
-            int fullTextFieldSizeLimit = fulltextConfiguration.fulltextFieldSizeLimit;
-            if (fullTextFieldSizeLimit != 0 && text.length() > fullTextFieldSizeLimit) {
-                if (log.isDebugEnabled()) {
-                    log.debug(String.format(
-                            "Fulltext extract of length: %s for indexName: %s of document: %s truncated to length: %s",
-                            text.length(), indexName, docId, fullTextFieldSizeLimit));
-                }
-                text = text.substring(0, fullTextFieldSizeLimit);
+            List<String> strings = new ArrayList<>();
+            for (Blob blob : blobsExtractor.getBlobs(document)) {
+                String string = blobsText.computeIfAbsent(blob, this::blobToText);
+                strings.add(string);
             }
-            indexesAndText.add(new IndexAndText(indexName, text));
-        }
-        if (!indexesAndText.isEmpty()) {
-            Work work = new FulltextUpdaterWork(repositoryName, docId, false, true, indexesAndText);
-            if (!fulltextConfiguration.fulltextSearchDisabled) {
-                WorkManager workManager = Framework.getService(WorkManager.class);
-                workManager.schedule(work, true);
-            } else {
-                ((FulltextUpdaterWork) work).updateWithSession(session);
+            String text = String.join("\n\n", strings);
+            text = limitStringSize(text, fulltextConfiguration.fulltextFieldSizeLimit);
```

Comment:
We talked about storing the full text in a transient store and raising an event.

Reply:
An event is already raised by `CoreSession.setDocumentSystemProp` but yes it doesn't have access to the value.

I'll change that method so that an event can get to the fulltext. Stay tuned.

Reply:
:+1: